src/responses.rs

```diff
@@ -314,19 +314,14 @@ pub struct FailureItem {
 /// 'pay' command
 #[derive(Debug, Clone, Deserialize, Serialize)]
 pub struct Pay {
-    pub id: u64,
     pub payment_hash: String,
     pub destination: String,
     pub msatoshi: u64,
     pub msatoshi_sent: u64,
-    pub created_at: u64,
+    pub created_at: f64,
```

Comment:
This should be an integer, it is a timestamp and from c-lightning view, it is a u32 value

https://github.com/ElementsProject/lightning/blob/a027b3a80cad8cd72c8629d44d687c48d9bfb181/plugins/pay.c#L1766

Reply:
Sorry for not getting back earlier. I just tested this with my local node again and it indeed gives me a float value :cry: 

```
{
   "destination": "<redacted>",                                      
   "payment_hash": "<redacted>                                       
   "created_at": 1623167235.903,
   "parts": 1,                                                                                                               
   "msatoshi": 11000,
   "amount_msat": "11000msat",
   "msatoshi_sent": 12003,
   "amount_sent_msat": "12003msat",
   "payment_preimage": "<redacted>",
   "status": "complete"
}
```

And before my changes it actually failed with

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Json(Error("invalid type: floating point `1623169487.421`, expected u64", line: 1, column: 227))', examples/pay.rs:17:54
```

Maybe that has to do with the JSON serialization? Technically JS has no integers, so maybe it's converted to float? If so we are in for some pain as also `msatoshi` could get so big that this happens and then serde would suddenly fail.

Reply:
Hi @sgeisler, no delay here.

>Maybe that has to do with the JSON serialization? Technically JS has no integers, so maybe it's converted to float? If so we are in for some pain as also msatoshi could get so big that this happens and then serde would suddenly fail.

If I'm not missing nothings the timestamp it is not a float, but I can miss some details. In any case, the JSON propriety must have one time, not multiple types, in this case, I think if there is some issue, the issue is inside the lightning. What version of lightning you are running?

Reply:
Ok, I dug into it: it's no floating point weirdness, but a [custom time serialization function](https://github.com/ElementsProject/lightning/blob/ade10e7fc4dacbb9d635b05152c7dc38c0896ce7/common/json.c#L1150) building something float-like. This is used to serialize some response struct [when a `libplugin-pay` payment finishes](https://github.com/ElementsProject/lightning/blob/76b8eb3afda4905dbeb059f88b2bbe994cbaa50a/plugins/libplugin-pay.c#L1866). My working theory is that a) this is saved somewhere and returned when querying the payment using `pay` and b) that the code you found is old and only there for compatibility reasons (most likely called when using `legacypay` in 0.10.0).

The question is what to make of all this? Build a custom parser that constructs a `SystemTime`? Just make it a float (potentially lossy, but probably not too much) because nobody uses it anyway?

Reply:
Also, I saw you are working on integration testing in #21, maybe you can simply reproduce this behavior instead of digging too deeply into the C code (which wasn't too much fun for me)? Alternatively running `lightning-cli pay <bolt11>` should show you the format of the `created_at` field.

Reply:
Hi @elsirion,

> Also, I saw you are working on integration testing in #21, maybe you can simply reproduce this behavior

Correct, and yes I started to work on this, but it is not too much easy because we need to create the network before. But, I'm working on it 😄 I already found some API changes.

>deeply into the C code

C is funny :-P (when you don't need to debug it :-) )

Sorry but today I had no time to look in to it

Reply:
I have the payload that proves that you have right.

```
      {
         "payment_hash": "ba4dd53019f0777ef82c649161cd7ee5ec0ae73ae98c83e957000b63d659f82f",
         "in_channel": "697341x2299x0",
         "out_channel": "688686x1433x0",
         "in_msatoshi": 18005467,
         "in_msat": "18005467msat",
         "out_msatoshi": 18004988,
         "out_msat": "18004988msat",
         "fee": 479,
         "fee_msat": "479msat",
         "status": "failed",
         "received_time": 1630926620.601,
         "resolved_time": 1630926758.306
      },

```

Thanks